src/expr/src/linear.rs

```diff
@@ -1881,6 +1882,132 @@ pub mod plan {
+                relation_desc,
+                mfp: mfp.cloned(),
+            }
+        }
+
+        /// Returns `false` if a chunk of data with the given statistics would
+        /// entirely be filtered out by this MFP once fetched.
+        ///
+        /// Returns `true` if we don't know.
+        pub fn should_fetch<S: PersistSourceDataStats>(&mut self, stats: &S) -> bool {
+            let mfp = match self.mfp.as_ref() {
+                Some(x) => x,
+                None => return true,
+            };
+            for (support, predicate) in mfp.mfp.predicates.iter() {
+                if mfp.mfp.input_arity < *support {
+                    // We'd need the result of map to evaluate this predicate.
+                    return true;
+                }
+                match self.provably_not_in_part(stats, predicate) {
+                    // The predicates are ANDed, so if any of them are provably
+                    // not in the part, we can confidently skip fetching it.
+                    Some(true) => return false,
+                    // TODO: There's probably some useful distinctions we could
+                    // pass back to persist for tracking: e.g. part definitely
+                    // is not needed vs part definitely might be needed vs we
+                    // don't know because some stats were missing vs we don't
+                    // know because the code doesn't yet handle some expression.
+                    Some(false) | None => continue,
+                }
+            }
+            // TODO: Do the same for self.map.{lower,upper}_bounds.
+            true
+        }
+
+        fn provably_not_in_part<S: PersistSourceDataStats>(
+            &self,
+            stats: &S,
+            expr: &MirScalarExpr,
+        ) -> Option<bool> {
+            let arena = RowArena::default();
+            // TODO: This is entirely just a placeholder impl so that we can get
+            // at least one query to work end-to-end. The real structure would
+            // certainly look different than this.
+            //
+            // It works by sniffing out expressions that (at a high level) look
+            // like `some_col {<,<=,>=,>} constant`. For `<`, if we substitute
+            // the minimal value in the part and if `some_col_min < constant` is
+            // _false_, then we know nothing else in the part would be true and
+            // we can skip fetching it. Analogous logic holds for the other
+            // comparisons, but with the max substituted for `{>=,>}`.
+            //
+            // The full generalization probably requires being able to reason
+            // about the range of outputs of functions given a restricted input
+            // domain, but... that seems hard? There's probably some nice middle
+            // ground.
+            if let MirScalarExpr::CallBinary { func, expr1, expr2 } = expr {
+                let col_idx = Self::one_column_reference(expr1)?;
```

Comment:
I'd say it's worth it to add handling for both orders before the feature gets in front of any users (maybe even Devex). There is a high chance that supporting only one of the expr orders would cause confusion whenever any users start trying this feature: somebody will probably write a query with the other expression order, and then waste time wondering why it's not working.

Btw. this made me think of one more question: How will the users know why the feature is not giving a speedup in a particular case? If there is no speedup, there are still several possibilities, e.g.:
- The expression-handling logic is failing on the particular predicate expression.
- The predicate is not filtering out enough data.
- The data is not distributed into chunks in a useful way, e.g., too few chunks, or every chunk containing at least a small portion of non-filtered data.
- Fetching all the data is almost as fast as fetching a small chunk of it, and some other part of the computation is slow for a reason that's not really tied to the amount of data, [e.g. some serializability issue](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1678892828001659?thread_ts=1678892667.121529&cid=C02PPB50ZHS).

Reply:
> I'd say it's worth it to add handling for both orders before the feature gets in front of any users (maybe even Devex). There is a high chance that supporting only one of the expr orders would cause confusion whenever any users start trying this feature: somebody will probably write a query with the other expression order, and then spend time wondering why it's not working.

yep, I agree 100%. I'd also want to add more datum types (currently only uint64) and unary cast funcs to the whitelist, among other things. I think to me it makes the most sense to grab some queries from devex/customers and then do a batch of whatever is needed to make them all work at once. my goal in this PR was just to get _at least one_ query working, as a way to keep the scope down and discussion focused more on the API boundary

> How will users know why the feature is not giving a speedup in a particular case?

this is a great question that I don't yet have an answer to. I left a TODO in the code in the PR about some distinctions that we might want to thread back up to persist for tracking (e.g. definitely skip vs definitely fetch vs we were missing stats vs we hit an `_` case in the code here). no idea how that information should make it back up to the user. for the short term, because this is behind a feature flag that we'll only turn on very selectively (not in prod), my plan is to put some of it in logs and dig it up as needed. that's obviously not a permanent solution, so I'm all ears if you (or anyone) has ideas. maybe something `EXPLAIN` based?

Reply:
> I think to me it makes the most sense to grab some queries from devex/customers and then do a batch of whatever is needed to make them all work at once. my goal in this PR was just to get at least one query working, as a way to keep the scope down and discussion focused more on the API boundary

Ok, makes sense, then let's merge this PR!

> maybe something EXPLAIN based?

Yes, maybe with some extra option. We have some options already, e.g., `EXPLAIN WITH(join_impls) SELECT ...` which prints info about join implementations. We could add something like
`EXPLAIN WITH(storage_filter_pushdown) SELECT ...`, which would print info about what predicate is pushed down, how many chunks are filtered out, etc. (@aalexandrov is the EXPLAIN expert.)

Reply:
very cool! any guesses how much time it might take for a quick initial impl of hooking the pushdown stuff into that? I bet it would be incredibly helpful, even in the early stages of getting eyes on this

Reply:
The `optimize_dataflow_filters` transform in `dataflow.rs` already pushes filters to sources!

https://github.com/MaterializeInc/materialize/blob/c926e6bee45a7386d3191951c8fd34e923a29ddb/src/transform/src/dataflow.rs#L314-L352

You need to figure out how the `SourceInstanceArguments` attached to `SourceInstanceDesc` are made available to `persist` in the rendering code (start from `DataflowDescription::source_imports`). There is a high chance that things still work and you just need to flip a switch somewhere. cc @vmarcos / @tokenrove who have a better overview of the rendering code.

Reply:
Filed a tracking issue for this: https://github.com/MaterializeInc/materialize/issues/18305